clang/include/clang/AST/ExprCXX.h

```diff
@@ -4902,6 +4902,98 @@ class BuiltinBitCastExpr final
   }
 };
 
+/// Represents a __builtin_num_fields expression.
+class SYCLBuiltinNumFieldsExpr : public Expr {
+  friend class ASTStmtReader;
+
+  SourceLocation Loc;
+  int64_t NumFields;
+  QualType SourceTy;
+
+public:
+  SYCLBuiltinNumFieldsExpr(SourceLocation Loc, int64_t NumFields,
+                           QualType SourceTy, QualType RetTy)
+      : Expr(SYCLBuiltinNumFieldsExprClass, RetTy, VK_RValue, OK_Ordinary),
+        Loc(Loc), NumFields(NumFields), SourceTy(SourceTy) {
+    setDependence(computeDependence(this));
+  }
+
+  explicit SYCLBuiltinNumFieldsExpr(EmptyShell Empty)
+      : Expr(SYCLBuiltinNumFieldsExprClass, Empty) {}
+
+  QualType getSourceType() const { return SourceTy; }
+  int64_t getNumFields() const {
+    assert(!isTypeDependent());
+    return NumFields;
+  }
+
+  SourceLocation getBeginLoc() const { return getLocation(); }
+  SourceLocation getEndLoc() const { return getLocation(); }
+
+  SourceLocation getLocation() const { return Loc; }
+  void setLocation(SourceLocation L) { Loc = L; }
+
+  child_range children() {
+    return child_range(child_iterator(), child_iterator());
+  }
+
+  const_child_range children() const {
+    return const_child_range(const_child_iterator(), const_child_iterator());
+  }
+
+  static bool classof(const Stmt *T) {
+    return T->getStmtClass() == SYCLBuiltinNumFieldsExprClass;
+  }
+};
+
+/// Represents a __builtin_field_type expression. This does not actually return
+/// the type itself (it would need to be a new type in the type system rather
+/// than an expression to do that), but instead is a valueless expression of
+/// the field's type. The expected usage is to use decltype to extract the
+/// actual type information. This is a hack. If we need to upstream this
+/// functionality, it should be done as a type in the type system instead.
```

Comment:
I'm not necessarily convinced we'd want to promote this to a type in the type system as I think this is a somewhat reasonable solution, but I definitely would have rathered this being able to return a type directly.

Reply:
I'm not certain I agree. Consider what the AST dump would look like if was modelled as an expression that resolves to a type as opposed to a type itself:
```
__builtin_field_type(Whatever, 0) x = 12;
```
We wouldn't want to see:
```
DeclStmt
 - SYCLBuiltinFieldTypeExpr
   - VarDecl (x)
     - IntegerLiteral (12)
```
I think if the expression generates a type, it needs to be a `Type` rather than an Expr directly (like `decltype`).

Reply:
Wouldn't that be what you would get if it was modeled as a type?  Currently that isn't particularly legal, since the builtin is a value-result.

Reply:
Nope, check out: https://godbolt.org/z/Tcnob7Mhd -- even though `decltype` is being given an expression, there's no expression node in the AST for it because `decltype(expr)` is still a way to spell a type.

Reply:
Okay, after some off-list discussion, I think we're on the same page here. It's fine for `__builtin_field_type` to return a value only so long as we can find a way to ensure it's an error if it's ever used in an evaluated context.